### ISSUE TITLE
Deadlock with multithreaded fork on OSX 10.12

### ISSUE THREAD

**Issue by alexcrichton** [jemalloc/jemalloc#895]
Over at rust-lang/rust we [recently upgraded](https://github.com/rust-lang/rust/pull/41861) jemalloc to 4.5.0, and I believe I've started noticing a deadlock locally in our tests when running on OSX 10.12. Specifically, the following program will deadlock:

```c
#include <assert.h>
#include <errno.h>
#include <pthread.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

static void work(void) {
  int child = fork();
  assert(child >= 0);
  if (child == 0) {
    char *args[] = {"sleep", "1000", NULL};
    execvp(args[0], args);
    printf("failed to spawn child: %s\n", strerror(errno));
    assert(0);
  }
  assert(kill(child, SIGKILL) == 0);
}

static void *worker(void *arg) {
  void *a = malloc(4);
  assert(a != NULL);
  work();
  free(a);
  return arg;
}

int main() {
  pthread_t child;
  assert(pthread_create(&child, NULL, worker, NULL) == 0);
  work();
  assert(pthread_join(child, NULL) == 0);
}
```

I built this via:

```
$ git clone https://github.com/jemalloc/jemalloc
$ cd jemalloc
$ git rev-parse HEAD
00869e39a334f3d869dfb9f8e651c2de3dded76f
$ ./autogen.sh
$ make -j10
# edit program above into `foo.c`
$ gcc foo.c lib/libjemalloc.dylib -o foo
$ DYLD_LIBRARY_PATH=lib ./foo
```

The deadlock isn't deterministic for me, but happens pretty frequently. It shouldn't take more than 10 or so tries to witness the deadlock. I'm personally running OSX 10.12.5 locally with what is I believe Xcode 8.3.2.

The backtrace at the time of deadlock looks like:

```
(lldb) bt all
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fffacf20f46 libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fffad00d6e5 libsystem_pthread.dylib`pthread_join + 425
    frame #2: 0x000000010e27dc0c foo`main + 124
    frame #3: 0x00007fffacdf2235 libdyld.dylib`start + 1
    frame #4: 0x00007fffacdf2235 libdyld.dylib`start + 1

  thread #2
    frame #0: 0x00007fffacf2131e libsystem_kernel.dylib`__ulock_wait + 10
    frame #1: 0x00007fffad004aff libsystem_platform.dylib`_os_ulock_wait + 25
    frame #2: 0x00007fffad0043d2 libsystem_platform.dylib`_os_unfair_lock_lock_slow + 130
    frame #3: 0x000000010e2b32f8 libjemalloc.2.dylib`je_malloc_mutex_lock_slow [inlined] malloc_mutex_lock_final(mutex=<unavailable>) at mutex.h:141 [opt]
    frame #4: 0x000000010e2b32f0 libjemalloc.2.dylib`je_malloc_mutex_lock_slow(mutex=0x000000010e530358) at mutex.c:83 [opt]
    frame #5: 0x000000010e29249c libjemalloc.2.dylib`je_arena_malloc_hard [inlined] malloc_mutex_lock at mutex.h:205 [opt]
    frame #6: 0x000000010e292480 libjemalloc.2.dylib`je_arena_malloc_hard [inlined] arena_malloc_small at arena.c:1495 [opt]
    frame #7: 0x000000010e292471 libjemalloc.2.dylib`je_arena_malloc_hard(tsdn=0x0000000000000000, arena=<unavailable>, size=<unavailable>, ind=30, zero=<unavailable>) at arena.c:1551 [opt]
    frame #8: 0x000000010e281b72 libjemalloc.2.dylib`a0ialloc [inlined] arena_malloc(tsdn=<unavailable>, arena=0x000000010e52a980, size=6144, slow_path=true) at arena_inlines_b.h:112 [opt]
    frame #9: 0x000000010e281b5e libjemalloc.2.dylib`a0ialloc [inlined] iallocztm(ind=30, slow_path=true) at jemalloc_internal_inlines_c.h:33 [opt]
    frame #10: 0x000000010e281b5e libjemalloc.2.dylib`a0ialloc(size=6144, zero=false, is_internal=true) at jemalloc.c:233 [opt]
    frame #11: 0x000000010e2833f8 libjemalloc.2.dylib`je_malloc at tsd_generic.h:73 [opt]
    frame #12: 0x000000010e2833d2 libjemalloc.2.dylib`je_malloc [inlined] tsd_get(init=true) at tsd_generic.h:140 [opt]
    frame #13: 0x000000010e2833d2 libjemalloc.2.dylib`je_malloc [inlined] tsd_fetch_impl(init=true) at tsd.h:254 [opt]
    frame #14: 0x000000010e2833d2 libjemalloc.2.dylib`je_malloc [inlined] tsd_fetch at tsd.h:272 [opt]
    frame #15: 0x000000010e2833d2 libjemalloc.2.dylib`je_malloc [inlined] imalloc at jemalloc.c:1907 [opt]
    frame #16: 0x000000010e28319e libjemalloc.2.dylib`je_malloc(size=4) at jemalloc.c:1944 [opt]
    frame #17: 0x00007fffacf74282 libsystem_malloc.dylib`malloc_zone_malloc + 107
    frame #18: 0x00007fffacf73200 libsystem_malloc.dylib`malloc + 24
    frame #19: 0x000000010e27dc7b foo`worker + 27
    frame #20: 0x00007fffad00b93b libsystem_pthread.dylib`_pthread_body + 180
    frame #21: 0x00007fffad00b887 libsystem_pthread.dylib`_pthread_start + 286
    frame #22: 0x00007fffad00b08d libsystem_pthread.dylib`thread_start + 13
```

This in turn led me to https://github.com/jemalloc/jemalloc/issues/843, although it's not precisely the same as this.  Perhaps that'd fix this issue though?

**Comment by jasone** [comment]
I vaguely recall (memory from implementing the runtime library for [Onyx](http://www.canonware.com/onyx/) circa 2002) that ```fork(2)``` is only reliable on macOS if invoked by the initial thread.  Has this been fixed?

**Comment by davidtgoldblatt** [comment]
It repros with jemalloc but not the system allocator, so it's at least jemalloc-involved.

**Comment by mhinz** [comment]
We have the same issue with Neovim. The [homebrew formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/neovim.rb) dynamically links jemalloc 5.0.1 which leads to more or less random deadlocks. When linking statically, everything seems to be fine.

[Reproduction steps + stacktrace](https://github.com/neovim/homebrew-neovim/issues/219#issuecomment-313707372).

**Comment by 0xABAB** [comment]
@alexcrichton This behavior is entirely in line with the documentation of the systems involved. 

If you have another deadlocking test program for which you can prove that it should work, feel free to open another issue.

This issue should be closed.

**Comment by davidtgoldblatt** [comment]
I mean, regardless of whether or not it's allowed by the OS X docs, we don't want to turn programs that work fine with the system allocator into ones that break with jemalloc.

I'll try to figure out the root cause here next week. Unfortunately a lot of these things boil down to "the libc implementation is hostile to malloc replacement", so a fix may be nontrivial or hurt fork performance.

**Comment by alexcrichton** [comment]
@davidtgoldblatt if the fix boils down do "don't replace the system malloc" that would be fine for us! We downgraded back to 4.0.0 temporarily while this is sorted out, but if that fixes everything we'd happily upgrade!

**Comment by 0xABAB** [comment]
@davidtgoldblatt Let's assume `jemalloc` is perfect. I already established that there is no reason to assume the test program should work. This implies that any action you can do to make `jemalloc` work with this broken program, will make `jemalloc` either bigger or it will make it slower. 

Both will cost Facebook (your employer), and everyone who writes correct programs (not sure if this includes Facebook) time and/or money to benefit a group of people who have a hard time writing programs (the Neovim community apparently). Is this part of some kind of extreme socialism trend that has hit San Francisco? 

Tell me, do they have a manifesto "Everyone's program must run"?

If not, why would you ever want to investigate this? It seems more useful to spend time on those other 86 issues, which at least (until I have read them) have a non-zero probability of being an actual issue.

**Comment by davidtgoldblatt** [comment]
Note that the test program demonstrated follows the posix rule of calling only async-signal-safe functions in the child, so if it's disallowed it's from something OS X-specific.

It seems like breakage doesn't happen in practice for OS X, so that the effective guarantee is the posix one, regardless of what's in the man pages (for the record: I haven't actually seen anything from OS X that says that this is disallowed, and Jason didn't remember a ton about it either; my current understanding is that the repro program should be guaranteed to work, and that this is a genuine jemalloc breakage).

**Comment by ilovezfs** [comment]
>s this part of some kind of extreme socialism trend that has hit San Francisco?

@0xABAB let's try to keep politics out of this. Thanks.

**Comment by ZyX-I** [comment]
@0xABAB You still have not described what the problem is exactly. And humans are unable to write perfect more or less complex something, which includes programs, documentation and bug reports. It is *you* who stated that program from the first comment is not supposed to work so it is *you* who needs to prove this, not @alexcrichton. Here are at least four sides who could’ve made an error: your stament may be wrong, OS X docs may be wrong, jemalloc code may be wrong or @alexcrichton program may be wrong. With the data you provided your comments look like “bug report is invalid, but I **cannot** write why [because I am a troll who wants people to spend time on uselessly rereading documentation]”.

And regardless of what exactly is wrong here what needs to be fixed is a separate issue: e.g. problem with @alexcrichton program may be too easy to miss or too common to be worth fixing on jemalloc side.

**Comment by 0xABAB** [comment]
@davidtgoldblatt Your message teaches me that I should get rid of `jemalloc` in all operational systems. Thanks!

@ilovezfs It's called sarcasm, not politics.

**Comment by ilovezfs** [comment]
@0xABAB either way, it's insulting to me and others who live in the Bay Area. So: please just stop.

**Comment by davidtgoldblatt** [comment]
Hi all,

From this point on, I'm going to start deleting comments that are not about the actual technical causes of deadlocks after a fork. Please feel free to open other issues to discuss bug prioritization, how to have polite technical disagreements, or the extent to which wanting multithreaded forking is a byproduct of San Francisco socialism.

**Comment by 0xABAB** [comment]
@ZyX-I I could write why, but I don't see the point in explaining things this easy to people who should know better.

I came into this topic, because it seemed fairly interesting, but after 5 seconds I already knew what was wrong. 

I think there is no room for democracy in proof (and reporting an issue). You believe that because @davidtgoldblatt believes something that it's true. That somehow, because someone else has an opinion in combination with a weak demonstration program, that I should feel obliged to show a proof why someone else is wrong. I think that just encourages laziness.

If @davidtgoldblatt wants to tell the story about how he wasted a week on something which was not a bug to his grandchildren, then that's his $DEITY given right.

Please tell me in a month that I was right about the program in the first comment. That would show sportsmanship.

**Comment by matthewbauer** [comment]
One thing I'm interested in is whether this is an OS X change or a jemalloc change that introduced this. @alexcrichton can you confirm that going back to 4.0.0 actually fixes this? Can we find the latest working version of jemalloc?

**Comment by alexcrichton** [comment]
@matthewbauer oddly enough if I compile anything pre-4.3.0 on my local mac (10.12) I keep running into #420. I believe that 4.0.0 works for ust in rust-lang/rust because we compile either on an older mac or... something? Unsure.

In any case I found that the first working commit I could find was a99e0fa and the next commit, 3f2b8d9c, ended up hitting the deadlock I found above. So I think that this is a consequence of 3f2b8d9c, although why 4.0.0 works for us in Rust already I'm... not sure!

**Comment by ilovezfs** [comment]
@alexcrichton yeah `malloc: *** malloc_zone_unregister() failed for 0xaddress` was a Sierra thing.

**Comment by davidtgoldblatt** [comment]
OK, I understand the hang.

On OS X, we rely on the zone machinery to call our prefork and postfork handlers.

In zone_force_unlock, we call jemalloc_postfork_child, reinitializing all our mutexes regardless of state. The comment says that this is required because the child can't unlock parent mutexes, even if they're locked by the forking thread. I find this claim dubious (it's certainly contrary to the intent of pthread_atfork), but maybe @jasone had a specific bug he was hitting.

To avoid unlocking the mutexes, the child reinitializes them in place, in both the parent and the child. In practice, this has the effect of resetting them to an unlocked state without doing the unlock action. In this case, it means clearing mutex state (arenas_lock in particular, but it could be any lock), but without waking the thread blocked on the mutex.

This works fine in the child (no other threads to wake anyway), but leaves the second thread hanging in the parent.

I'm inclined to just fix this by switching the jemalloc_postfork_child() call in zone_force_unlock() to jemalloc_postfork_parent() (which fixes the hang at least, and, AFAICT introduces no issues in the child), but I want to double-check that, if the mutex re-init thing is ever necessary, it's at least not necessary on OS X.
